### DISCUSSION TITLE
How can I assert that a download was triggered with ui.download?

### DISCUSSION THREAD

**Discussion by marcuslimdw** [zauberzeug/nicegui#3686] [Q&A] [answered]
### Question

My UI contains a `ui.download` element. I want to assert that this element was in fact triggered with specific content. How can I do so?

Right now, it seems like that's not possible. If so, I'm not sure if this is the desired approach, but one way I can see to do this in tests is:

1. Create a separate `TestingOutbox` implementation that, in addition to its real functionality, tracks download events and stores them separately
2. Use `TestingOutbox` instead of the normal `Outbox` in `testing.User`
3. Add a `assert_downloaded` method on `testing.User` that will check this new attribute on `TestingOutbox`

**Comment by rodja** [comment]
Yes you could create a special outbox for testing purposes. But it can also be archived with the existing tools. Given a page like this

```py
    @ui.page('/')
    def page():
        ui.button('Download', on_click=lambda: ui.download(b'Hello', filename='hello.txt'))
```

You could check the download like this:

```py
    await user.open('/')
    user.find('Download').click()
    assert user.client
    messages = list(user.client.outbox.messages)
    assert len(messages) == 1
    target_id, message_type, data = messages[0]
    assert message_type == 'download'
    assert data['src'] == b'Hello'
    assert data['filename'] == 'hello.txt'
```

**Reply by marcuslimdw** [reply]
When I tested it, the outbox was empty, but I realise now that it was due to some error in my application code - I thought that the outbox had been cleared by some internal operation.

In that case, would you accept a PR to add a new method to `User` (something like the aforementioned `assert_downloaded`) that wraps the above process to make it more ergonomic to test?

**Reply by marcuslimdw** [reply]
After some experimentation, I think that that will not work.

The problem here is that the `Outbox` contains the download message *only if* it is inspected right when the download is triggered.

In my specific case, this is what happens:

1. The user performs an action
2. In the background, some preparations are made
3. Once step 2 is done, a download is triggered
4. Some other stuff happens
5. The "act" phase of the test ends and I perform assertions

Because the `Outbox` instance on `User` works just like a normal outbox, by the time we reach step 5, it has handled all the pending messages, including the download message, and is now empty.

This is why, based on my current understanding, there is a need for:

1. Some functionality to store all download (and perhaps other types thereof) messages that have been seen by this user
2. An API to perform assertions on these stored messages

**Comment by rodja** [comment, marked as answer]
I created https://github.com/zauberzeug/nicegui/pull/3689 to provide such an API. What do you think?
